python/ray/data/dataset_pipeline.py

```diff
@@ -652,8 +655,14 @@ def iter_datasets(self) -> Iterator[Dataset[T]]:
         if self._executed[0]:
             raise RuntimeError("Pipeline cannot be read multiple times.")
         self._executed[0] = True
-        self._optimize_stages()
-        return PipelineExecutor(self)
+        iter = None
+        if self._first_dataset:
+            iter = itertools.chain([self._first_dataset], self._dataset_iter)
+        else:
+            self._peek()
+            iter = itertools.chain([self._first_dataset], self._dataset_iter)
+        self._dataset_iter = None
+        return iter
```

Comment:
Even simpler, I think this should work:
```suggestion
        if self._first_dataset is None:
            self._peek()
        iter = itertools.chain([self._first_dataset], self._dataset_iter)
        self._dataset_iter = None
        self._first_dataset = None
        return iter
```
We also need to unlink the `._first_dataset` here to keep that reference from hanging around.

Reply:
We need _first_dataset for the lifetime of DatasetPipeline to answer schema() call.
An alternative is to keep the schema cached not the dataset, but that will have to force read the schema if it's missing.

Reply:
Ensuring that the first `Dataset` is released after execution/consumption of the full pipeline starts (i.e. `.iter_datasets()` is called) is a strong requirement here, since the core `DatasetPipeline` semantics are that datasets are aggressively cleaned up as soon as the pipeline consumer has moved past the dataset. All `DatasetPipeline` use cases rely on these semantics, which is why a `DatasetPipeline` can only be read once in general.

There are two questions here:
1. Do we need to support `pipe.schema()` after the full pipeline has already started being consumed?
2. If so, how would we support it?

Given that users typically query the schema _before_ starting to iterate over the pipeline, the answer to (1) may be no. I can't immediately think of why the user would want to poke at the schema during execution;  `.iter_datasets()` is generally the "sink" of the pipeline. And if they really wanted to poke at the schema, they could get the most up-to-date schema by inspecting the schema for the most recently yielded dataset from `.iter_datasets()`.

Assuming that the answer to (1) is yes, two options immediately come to mind for supporting it:
1. We always peek when starting full pipeline consumption and cache the schema. This is simple, but is a less general solution if we want to calculate other things off of the peeks (e.g. other metadata).
2. We always keep a reference to the last computed dataset. This initially sounds good, but this will prevent us from being able to release the last dataset in the pipeline while we're computing the next dataset, which is a nonstarter.

We could try out the first solution or we could have `.schema()` throw an error that you can't fetch the schema of a pipeline that's currently being consumed. I'm leaning towards the latter I think.

Reply:
Basically, we'd allow this:
```python
pipe = ds.read._parquet(...).repeat()
print(pipe.schema())
for ds in pipe.iter_datasets():
   ...
```

but we wouldn't allow this

```python
pipe = ds.read._parquet(...).repeat()
iter = pipe.iter_datsets()
print(pipe.schema())
for ds in iter:
   ...
```

And the user would get an error message telling them to inspect the schema before calling `.iter_datasets()` (or other higher-level consuming APIs).

Reply:
If removing dataset after iteration is a hard requirement, then the path will likely be taking the alternative (cache schema), in that case we need to unconditionally fetch in the schema() call.

If there are those tricky conditions on when user can call schema(), I think the interface is becoming more difficult to use and also less useful. A more proper way is I think just removing schema() from DatasetPipeline and direct users to access it from Dataset APIs.

Reply:
The peek-based schema functionality, where we cache the first dataset until execution starts, will support most use cases in which a user wants to get the schema of a pipeline, and I think this will be useful and will eliminate some redundant computation users are already doing in order to read the schema on a pre-pipeline dataset (which won't even be the most correct schema if there are post-pipeline operations).

But I do agree that the interface/semantics in which you can only fetch the schema before starting consumption is pretty displeasing...

Reply:
I think it should be renamed (or add a new one if in-place renaming is not allowed) to something more accurate, like peek_schema(), rather than schema().

Reply:
That sounds good to me, that does a much better job of conveying the semantics!

Reply:
> We always keep a reference to the last computed dataset. This initially sounds good, but this will prevent us from being able to release the last dataset in the pipeline while we're computing the next dataset, which is a nonstarter.

Good point!